### PR TITLE
Add df_summary function

### PR DESCRIPTION
**Pull request by kyleam**
Re: #814

---

~~This has a trivial conflict with #816~~

### CODE REVIEW

File: pymc3/stats.py [hunk 1]
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
...
+
+    Returns
+    -------
+    Data frame with summary statistics for each variable
+    """
+    if varnames is None:
+        varnames = trace.varnames
+
+    if stat_funcs is None:
+        stat_funcs = [lambda x: pd.Series(np.mean(x, 0), name='mean_'),
```
Comment (twiecki) [inline pymc3/stats.py]: I suppose this is done so as not to collide with the `mean()` method? Note that pandas will would still call `mean()` if you do `df.mean()` even with such a column present, but evaluate the column if you did `df.mean`.

File: pymc3/stats.py [hunk 2]
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
...
+           column of the passed in array and the Series name is the
+           name of the statistic
+        2) or a Pandas DataFrame instance where the rows correspond to
+           the columns of the passed in array and the column names are
+           the names of the statistics.
+    alpha : float
+        The alpha level for generating posterior intervals. Defaults
+        to 0.05. This is only meaningful when `stat_funcs` is None.
+    batches : int
+      Batch size for calculating standard deviation for
```
Comment (twiecki) [inline pymc3/stats.py]: Bad indent.

File: pymc3/stats.py [hunk 3]
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
...
+      non-independent samples. Defaults to 100. This is only
+      meaningful if `stat_funcs` when None.
```
Comment (twiecki) [inline pymc3/stats.py]: when->is?

File: pymc3/stats.py [hunk 4]
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
...
+
+    See also
+    --------
+    summary : Generate a pretty-printed summary of a trace.
+
+    Returns
+    -------
+    Data frame with summary statistics for each variable
```
Comment (twiecki) [inline pymc3/stats.py]: Data frame -> pandas.DataFrame

**Comment by twiecki** [comment]
See minor comments but otherwise LGTM.

File: pymc3/stats.py (hunk 2 above)
Comment (kyleam) [inline pymc3/stats.py]: Thanks for the comments.  Will update.

File: pymc3/stats.py (hunk 1 above)
Comment (kyleam) [inline pymc3/stats.py]: > I suppose this is done so as not to collide with the `mean()` method?

Yep.

> Note that pandas will would still call `mean()` if you do `df.mean()`
> even with such a column present, but evaluate the column if you did
> `df.mean`.

Is that a feature of a recent pandas?  With the version of pandas I have
at hand (0.16.2), the method call will still work, but trying to select
the column with `.mean` does not (you instead have to use `[mean]`).

**Comment by kyleam** [comment]
Thanks for the comments. Will update.

File: pymc3/stats.py (hunk 1 above)
Comment (twiecki) [inline pymc3/stats.py]: You're right, I got myself confused. I would still argue that `mean` is the better name. For example, `.describe()` also creates `mean`.

Comment (kyleam) [inline pymc3/stats.py]: Good point about `describe`.   `mean` it is.

File: pymc3/stats.py [hunk 5]
```diff
@@ -232,6 +237,71 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5)):
...
+        A list of functions used to calculate statistics. By default,
+        the mean, standard deviation, simulation standard error, and
+        highest posterior density intervals are included.
+
+        The functions will be given one argument, the samples for a
+        variable as a 2 dimensional array, where the first axis
+        correspond to sampling iterations and the second axis
+        represent the flattened variable (e.g., x__0, x__1,...). It
+        should return either
+        1) a Pandas Series instance where each value corresponds to a
```
Comment (fonnesbeck) [inline pymc3/stats.py]: Function passing via a pandas data structure ("pandas" as a name is not capitalized, FWIW) is not entirely clear to me. Maybe add an example?

**Comment by fonnesbeck** [comment]
Looks great to me. This is a great idea.

File: pymc3/stats.py (hunk 5 above)
Comment (kyleam) [inline pymc3/stats.py]: > Maybe add an example?

Will do.

> "pandas" as a name is not capitalized, FWIW

Thanks.

**Comment by kyleam** [comment]
Changes since v1:
- Updated for comments.
- Added `extend` argument to `df_summary`.
- Exposed stats.py's `__all__` in top-level namespace.
- Tried to improve `stat_funcs` argument description.

**Comment by twiecki** [comment]
:+1: This is great stuff!

**Pull request merged**